spec.md

```diff
@@ -46,43 +49,55 @@ For relevant details and a history leading up to this specification, please see
-- V2 Client (Consumer) requirements
+### Artifact Verification
 
-### Future
+A container engine would like to run verified image named "library/ubuntu", with the tag "latest".
+The engine contacts the registry, requesting the manifest for "library/ubuntu:latest".
+An untrusted registry returns a manifest.
+Before proceeding to download the individual layers, the engine verifies the manifest's signature, ensuring that the content was produced from a trusted source and no tampering has occurred.
+After each layer is downloaded, the engine verifies the digest of the layer, ensuring that the content matches that specified by the manifest.
 
-The following is an incomplete list of features, discussed during the process of cutting this specification, which MAY be out of the scope of this specification, MAY be the purview of another specification, or MAY be deferred to a future version:
+### Resumable Push
 
-- Authentication and authorization support: While authentication and authorization support will influence this specification, those details MAY be left to a future specification. However, relevant header definitions and error codes are present to provide an indication of what a client MAY encounter.
-- Immutable image references
-- Multiple architecture support
-- Migration from v2compatibility representation
+Company X's build servers lose connectivity to a distribution endpoint before completing an artifact layer transfer.
+After connectivity returns, the build server attempts to re-upload the artifact.
+The registry notifies the build server that the upload has already been partially attempted.
+The build server responds by only sending the remaining data to complete the artifact file.
 
-## Conformance
-TODO: add general text about artifact validation requirements
+### Resumable Pull
 
-### Notational Conventions
+Company X is having more connectivity problems but this time in their deployment datacenter.
+When downloading an artifact, the connection is interrupted before completion.
+The client keeps the partial data and uses http `Range` requests to avoid downloading repeated data.
 
-The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" are to be interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119) (Bradner, S., "Key words for use in RFCs to Indicate Requirement Levels", BCP 14, RFC 2119, March 1997).
+### Layer Upload De-duplication
+
+Company Y's build system creates two identical layers from build processes A and B.
+Build process A completes uploading the layer before B.
+When process B attempts to upload the layer, the registry indicates that its not necessary because the layer is already known.
+
+If process A and B upload the same layer at the same time, both operations will proceed and the first to complete will be stored in the registry (Note: we MAY modify this to prevent dogpile with some locking mechanism).
```

Comment:
This note part seems odd to me. Maybe mention that the registry may securely only store one copy of the layer since the computed digests match, even though both upload are accepted.

Reply:
I've added some language tp reflect your comment. I think we should revisit the use-cases section